File: workers/research/src/black_book_research/discovery/campaign.py
"""Discovery campaign boundaries and budget enforcement (BB-039)."""

from __future__ import annotations

from black_book_research.adapters.types import AdapterCandidateRecord

from .geography import extract_geographic_hints, geographic_hint_within_countries
from .types import DiscoveryCampaignBoundaries, DiscoveryCampaignBudget


def assert_campaign_budget_valid(budget: DiscoveryCampaignBudget) -> None:
    if budget.max_candidates < 1:
        raise ValueError("Campaign max_candidates must be at least 1")
    if budget.max_quarantined < 0:
        raise ValueError("Campaign max_quarantined must be non-negative")
    if budget.max_dead_letter < 0:
        raise ValueError("Campaign max_dead_letter must be non-negative")
    if budget.max_retries_per_candidate < 0:
        raise ValueError("Campaign max_retries_per_candidate must be non-negative")


def record_within_campaign_boundaries(
    record: AdapterCandidateRecord,
    boundaries: DiscoveryCampaignBoundaries,
) -> bool:
    if boundaries.adapter_ids:
        if record.provenance.adapter_id not in boundaries.adapter_ids:
            return False
    hints = extract_geographic_hints(record)
    return geographic_hint_within_countries(hints, boundaries.countries)

File: packages/domain/src/discovery/campaign.ts
/**
 * Discovery campaign boundaries and budget enforcement (BB-039).
 */
import type { AdapterCandidateRecord } from '../adapters/types.js';
import type {
  DiscoveryCampaignBoundaries,
  DiscoveryCampaignBudget,
  DiscoveryCampaignConfig,
} from './types.js';
import { geographicHintWithinCountries, extractGeographicHints } from './geography.js';

export function assertCampaignBudgetValid(budget: DiscoveryCampaignBudget): void {
  if (budget.maxCandidates < 1) {
    throw new Error('Campaign maxCandidates must be at least 1');
  }
  if (budget.maxQuarantined < 0) {
    throw new Error('Campaign maxQuarantined must be non-negative');
  }
  if (budget.maxDeadLetter < 0) {
    throw new Error('Campaign maxDeadLetter must be non-negative');
  }
  if (budget.maxRetriesPerCandidate < 0) {
    throw new Error('Campaign maxRetriesPerCandidate must be non-negative');
  }
}

export function assertCampaignBoundariesValid(boundaries: DiscoveryCampaignBoundaries): void {
  if (boundaries.countries.length === 0) {
    throw new Error('Campaign boundaries require at least one country or global');
  }
}

export function createDiscoveryCampaignConfig(
  input: DiscoveryCampaignConfig,
): DiscoveryCampaignConfig {
  assertCampaignBudgetValid(input.budget);
  assertCampaignBoundariesValid(input.boundaries);
  return input;
}

export type CampaignBudgetSnapshot = {
  readonly accepted: number;
  readonly quarantined: number;
  readonly deadLetter: number;
  readonly totalProcessed: number;
};

export function isWithinCandidateBudget(
  snapshot: CampaignBudgetSnapshot,
  budget: DiscoveryCampaignBudget,
): boolean {
  return snapshot.totalProcessed < budget.maxCandidates;
}

export function isWithinQuarantineBudget(
  snapshot: CampaignBudgetSnapshot,
  budget: DiscoveryCampaignBudget,
): boolean {
  return snapshot.quarantined <= budget.maxQuarantined;
}

export function isWithinDeadLetterBudget(
  snapshot: CampaignBudgetSnapshot,
  budget: DiscoveryCampaignBudget,
): boolean {
  return snapshot.deadLetter <= budget.maxDeadLetter;
}

export function isWithinCampaignBudget(
  snapshot: CampaignBudgetSnapshot,
  budget: DiscoveryCampaignBudget,
): boolean {
  return (
    isWithinCandidateBudget(snapshot, budget) &&
    isWithinQuarantineBudget(snapshot, budget) &&
    isWithinDeadLetterBudget(snapshot, budget)
  );
}

export function recordWithinCampaignBoundaries(
  record: AdapterCandidateRecord,
  boundaries: DiscoveryCampaignBoundaries,
): boolean {
  if (boundaries.adapterIds && boundaries.adapterIds.length > 0) {
    if (!boundaries.adapterIds.includes(record.provenance.adapterId)) {
      return false;
    }
  }

  const hints = extractGeographicHints(record);
  return geographicHintWithinCountries(hints, boundaries.countries);
}
